### PR TITLE
fix: show more errors in `SignOrExecuteForm`

### LINKED ISSUE #874: [Tx execution]Tx most likely fail message is missed in some cases
## Environment
 - Browser: Chrome
 - Wallet: MetaMask
 - Chain: Ethereum mainnet

## Steps to reproduce
 1. Go to https://max-fee.web-core.pages.dev/gno:0x9DF58F04F3e914ED8D972146401C85313A53FecA/transactions/queue
 2. try to execute the tx with the nonce =40
Current result: "Tx most likely fail" message should be displayed if we can detect it without the simulation 
## Expected result
image

### PR DESCRIPTION
**Pull request by iamacook**
## What it solves

Resolves #874

## How this PR fixes it

Errors are now shown for: wrong chain, non-proposal, sending to self, gas-related errors and submission errors.

## How to test it

- Connect to wrong chain and observe the "Your wallet is connected to the wrong chain." error.
- Try creating a transaction as a non-owner and observe the "You are currently not an owner of this Safe and won't be able to submit this transaction." error.
- Connect to a nested Safe and send funds to the parents. Observe the "Cannot execute a transaction from the Safe itself, please connect a different account." (This was present in `safe-react` but I am of the opinion that we should remove this, or extend it to all instances of owners?)
- Attempt to execute the `swapOwner` on `gno:0x9DF58F04F3e914ED8D972146401C85313A53FecA` and observe the `safeTxGas`-related error.

Note: gas-related errors change wording depending on if they are related to creation or execution:

- "This transaction will most likely fail. To save gas costs, avoid creating the transaction."
- "This transaction will most likely fail. To save gas costs, reject this transaction."

## Screenshots

image

image

### CODE REVIEW

File: src/components/tx/SignOrExecuteForm/index.tsx [hunk 1]
```diff
@@ -79,14 +81,17 @@ const SignOrExecuteForm = ({
   // Estimate gas limit
   const { gasLimit, gasLimitError, gasLimitLoading } = useGasLimit(willExecute ? tx : undefined)
 
+  // Check if transaction will fail
+  const { isValidExecutionError, isValidExecutionLoading } = useIsValidExecution(willExecute ? tx : undefined)
```
Comment (iamacook) [inline src/components/tx/SignOrExecuteForm/index.tsx:85]: Spending limits currently show an error as they don't call `execTransaction`.

File: src/hooks/useIsValidExecution.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  const safeSDK = useSafeSDK()
+
+  const [isValidExecution, isValidExecutionError, isValidExecutionLoading] = useAsync(() => {
+    if (!safeTx || !safeSDK) {
+      return
+    }
+
+    return safeSDK
+      .getContractManager()
+      .safeContract.execTransaction(safeTx)
```
Comment (iamacook) [inline src/hooks/useIsValidExecution.ts:22]: Spending limits currently show an error as they don't call `execTransaction`.

File: src/components/tx/SignOrExecuteForm/index.tsx [hunk 3]
```diff
@@ -196,23 +202,26 @@ const SignOrExecuteForm = ({
...
+              ? 'To save gas costs, avoid creating the transaction.'
+              : 'To save gas costs, reject this transaction.'}
           </ErrorMessage>
-        )}
-
-        {submitError && (
+        ) : submitError ? (
           <ErrorMessage error={submitError}>Error submitting the transaction. Please try again.</ErrorMessage>
-        )}
+        ) : null}
```
Comment (iamacook) [inline src/components/tx/SignOrExecuteForm/index.tsx:224]: Still needs refactoring.

File: src/hooks/useIsValidExecution.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import type { SafeTransaction } from '@gnosis.pm/safe-core-sdk-types'
+
+import { useSafeSDK } from './coreSDK/safeCoreSDK'
+import useAsync from './useAsync'
+
+const useIsValidExecution = (
+  safeTx?: SafeTransaction,
+): {
+  isValidExecution?: boolean
+  isValidExecutionError?: Error
```
Comment (katspaugh) [inline src/hooks/useIsValidExecution.ts:10]: We start only booleans with `is`.

File: src/components/tx/SignOrExecuteForm/index.tsx [hunk 5]
```diff
@@ -170,7 +178,8 @@ const SignOrExecuteForm = ({
 
   const cannotPropose = !isOwner && !onlyExecute // Can't sign or create a tx if not an owner
   const submitDisabled = !isSubmittable || isEstimating || !tx || disableSubmit || isWrongChain || cannotPropose
-  const error = props.error || (willExecute && gasLimitError)
+  const error = props.error || (willExecute ? gasLimitError || executionValidationError : undefined)
+  const isSendingToSelf = wallet ? sameString(wallet?.address, safeAddress) : false
```
Comment (katspaugh) [inline src/components/tx/SignOrExecuteForm/index.tsx:182]: A misnomer, it's not sending to self, it's the Safe executing its own tx.
Sending to self works fine.

File: src/hooks/useIsValidExecution.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    if (!safeTx || !wallet?.address || !gasLimit) {
+      return
+    }
+
+    const { contract } = getSpecificGnosisSafeContractInstance(safe)
+
+    console.log('======================================== contract', contract)
+
+    try {
+      const test = await contract.callStatic.execTransaction(
```
Comment (iamacook) [inline src/hooks/useIsValidExecution.ts:37]: Leftover variable name from testing. I'll revert it when I push fixes for the test.

Comment (iamacook) [inline src/hooks/useIsValidExecution.ts:37]: Will be reverted when I push the fix for the test.

File: src/hooks/__tests__/useIsValidExecution.test.ts [hunk 7]
```diff
@@ -102,18 +115,22 @@ describe('useIsValidExecution', () => {
       },
     })
 
-    const { result } = renderHook(() => useIsValidExecution(createSafeTx(), BigNumber.from(1_000)))
+    const { result } = renderHook(() => useIsValidExecution(mockTx, mockGas))
 
-    const { isValidExecution, executionValidationError, isValidExecutionLoading } = result.current
+    var { isValidExecution, executionValidationError, isValidExecutionLoading } = result.current
```
Comment (katspaugh) [inline src/hooks/__tests__/useIsValidExecution.test.ts:120]: `var`? :)

**Review by katspaugh — APPROVED** [review]
LGTM! 👍

File: src/hooks/__tests__/useIsValidExecution.test.ts (hunk 7 above)
Comment (iamacook) [reply src/hooks/__tests__/useIsValidExecution.test.ts:120]: I was having reassignment issues and found [this example in the `useLoadChains` test](https://github.com/safe-global/web-core/blob/dev/src/hooks/__tests__/useLoadChains.test.ts#L39-L51) that helped.

A fresh head has helped though, I've adjusted it to use `let` instead.

**Comment by francovenica** [comment]
LGTM:
image
image
image
image

**Pull request merged**